sklearn/manifold/t_sne.py

```diff
@@ -712,7 +712,7 @@ def _fit(self, X, skip_num_points=0):
 
             # Find the nearest neighbors for every point
             neighbors_method = 'ball_tree'
-            if (self.metric == 'precomputed'):
+            if (self.metric == 'precomputed') or (self.metric == "cosine"):
```

Comment:
I don't get why this logic is here at all. I think we should always be using `neighbors_method='auto'`, which will use `ball_tree` when it can. Ping @tomMoral (who should have his name in the tsne source code)?

Reply:
I agree. Shell I make a new pull request?